[Feat] 네비게이션 바 구현

daminoworld opened the pull request:
## 🥕 관련 이슈

<!-- 관련있는 이슈 번호(#000)을 적어주세요. -->
- relative #1 

## 🥕 구현/변경 사항 및 이유

<!-- 구현/변경한 내용과 그 이유를 적어주세요. -->
- 네비게이션 바, 바 버튼 아이템 구현
   - 바 버튼 아이템 사이 간격 조절
   - extension을 통해 네비게이션 바 아이템 세팅 기능 구현
   - 바 버튼 아이템에 들어갈 아이콘 이미지 교체 (이미지 자체 여백이 있어 간격조절이 안되는 문제 해결)
   
## 🥕 PR Point

<!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
- 네비게이션 바 구현 및 간격 조절을 위해 코드로 구현했는데 더 효율적으로 짤 수 있는 방법이 있는지 궁금합니다..ㅎ 
  아니면 역시 네비게이션 바 hidden 하고 UIVIew로 구현하는게 더 효율적일까요?

## 🥕 참고 사항

<!-- 참고할 사항(+스크린샷)이 있다면 적어주세요. -->

daminoworld committed:
[Feat] 네비게이션 바 구현

- 네비게이션 바, 바 아이템 구현
- 바 아이템 사이 간격 조절

daminoworld committed:
[Feat] CollectionView를 이용한 아이템 리스트 구현

- CollectionView 및 Cell xib를 이용해 아이템 리스트 구현

daminoworld committed:
[Chore] 일부 코드 수정

- Create button IBAction 추가
- 컨벤션 어긋 나는 일부 코드 수정

noah0316 commented on Danggn/Resource/Assets.xcassets/ButtonIcons/icon_check_selected.imageset/Contents.json:
```diff
@@ -1,21 +1,26 @@
   "images" : [
     {
-      "filename" : "and_icon_menu.png",
+      "filename" : "Property 1=ios_check_on.png",
```
filename에 공백이 들어가 있으면 파일을 제대로 인식을 못하는 경우가 있습니다!!

daminoworld replied:
저 json 파일네임은 어떻게 변경을 해줘야 할까요..?

noah0316 commented on Danggn/Application/AppDelegate.swift:
```diff
@@ -10,8 +10,6 @@ import UIKit
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
-
```
혹시 lint에 AppDelegate가 걸렸나요? 
018b78cef878097144b65d75998961fffdf33930 여기에서 제외해두었는데! 혹시나해서 여쭤봅니다:)

cchanmi replied:
여기를 눌러 보면 웹페이지를 찾지 못하였다고 뜨네요...!

daminoworld replied:
AppDelegate가 lint에 걸리진 않았던것 같은데 어디 이상이 있나요?

noah0316 replied:
@daminoworld 
AppDelegate 파일 수정내역이 보여서 여쭤보았습니다:)

noah0316 replied:
@cchanmi 
링크 수정했습니다. 감사합니다:)

noah0316 commented on Danggn/Global/Extension/UINavigationItem+Extension.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension UINavigationItem {
+    
+    func setRightButtonUI(_ target: Any?, imageName: UIImage?) -> UIBarButtonItem {
```
`setRightButtonUI` 메소드 내부에서 `target`이라는 parameter가 사용되지 않고 있는데 따로 적어두신 이유가 있을까요? 😃 버튼에 action을 붙여주려고 하신 것 이겠지요??

daminoworld replied:
나주에 action이 들어갈 수 있으니까 그냥 넣었던것 같은데 당장 쓸 일이 없으면 그냥 빼는게 나을까요?ㅎㅎ

noah0316 replied:
이 함수(`setRightButtonUI(_: imageName:)`)에서 action을 같이 설정해주려면 
함수 인자에 action도 들어가있어야 하지않을까요? 
`button.addTarget(target, action: action, for: .touchUpInside)`와 같은 식으로 사용하니까요!
![Screen Shot 2022-05-30 at 0 43 57](https://user-images.githubusercontent.com/63908856/170878273-7c3cc643-b1d4-464f-ac67-0a648eedbfe0.png)

함수명이 setRightButtonUI인데, action까지 같이 설정해주는 것은 함수 네이밍 혹은 책임분리에 대해 생각해볼만한 지점인 것 같습니다:)

daminoworld replied:
음 그렇네요 애초에 setRightButtonUI니까 UI 관련설정만 해주는게 나은것 같습니다. target 인자를 빼는것으로 수정하였습니다!

daminoworld committed:
[Fix] 충돌 해결시도 커밋

- ItemDetailViewController, Stroryboard 추가
- ListItemViewController 내 다른 코드 추가

- 통신 관련 파일 추가 (진행 중)

daminoworld committed:
Merge branch 'develop' into feature/1-list-item

noah0316 commented on Danggn/Screen/Home/Controller/ListItemCollectionViewCell.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    func setBorder() {
+        listItemBorderView.clipsToBounds = true
+        listItemBorderView.layer.cornerRadius = 10
+        listItemBorderView.layer.maskedCorners = CACornerMask(arrayLiteral: .layerMinXMinYCorner, .layerMaxXMinYCorner)
```
listItemBorderView를 확인해보니 이 친구는 TableView의 Seperator와 같은 역할을 하는 친구 같은데 따로 하단 모서리 부분만 둥글게 만드신 이유가있는지 궁금합니다!!
<img width="297" alt="Screen Shot 2022-05-28 at 9 42 39" src="https://user-images.githubusercontent.com/63908856/170803015-83ee665f-4874-4e3d-8af6-5557133daa4d.png">

daminoworld replied:
디자인을 자세히 보니 아이템별 구분선 부분에 라운드 처리가 되어있어 그 부분을 구현할 코드를 찾다가 쓴건데 아무래도 코드를 가져오다보니 실수로 아래부분만 적용되는 코드로 써버렸네요,,ㅎㅎ 좋은 지적 감사합니다! 윗부분도 적용되도록 수정하였습니다!

noah0316 commented on Danggn/Screen/Home/Controller/ListItemViewController.swift:
```diff
@@ -8,10 +8,105 @@
+        registerCell()
+    }
+    
+    @IBAction func itemCreateButtondidTap(_ sender: Any) {
```
Button이 호출을 해주는 메소드이니 UIButton으로 명시화를 해주는건 어떨까요?
```suggestion
    @IBAction func itemCreateButtondidTap(_ sender: UIButton) {
```

daminoworld replied:
이부분 수정하겠습니다 감사합니다!

daminoworld replied:
이제야 확인한건데 didTap에서 DidTap이라고 적었어야되는데  실수로 didTap이라고 썼는데 lint에서는 감지가 안된건가요?

noah0316 replied:
SwiftLint에서는 아쉽게도 특정 함수 네이밍 까지는 파싱해주지는 않습니다
우리 프로젝트에 대한 lint rule은 같이 회의에서 정한 것 처럼 
https://github.com/THE-SOPT-Joint-Seminar-Danggn/Daangn-iOS#swiftlint-rule 에 명시되어있습니다:)

noah0316 replied:
대소문자와 관련한 rule은 https://realm.github.io/SwiftLint/identifier_name.html 해당 룰이 그나마 가장 근접한 것 같네요:)

noah0316 commented on Danggn/Screen/Home/Controller/ListItemViewController.swift:
```diff
@@ -8,10 +8,105 @@
+extension ListItemViewController {
+    func setNavigationUI() {
+        
+        let navigationBar = UINavigationBarAppearance()
```
```suggestion
        let navigationBarAppearance = UINavigationBarAppearance()
```
navigationBarApperance가 더 좋은 네이밍이지 않을까 싶습니다!

30번 라인에서 
```swift
self.navigationController?.navigationBar.standardAppearance = navigationBarAppearance
```
로 되어있는게 혼동의 여지없이 더 잘 읽히는 것 같습닡다!!😀

daminoworld replied:
좋은것 같습니다! 해당사항 반영해서 수정하겠습니다! 감사합니다~!

noah0316 commented on Danggn/Screen/Home/Controller/ListItemViewController.swift:
```diff
@@ -8,10 +8,105 @@
+    
+    // CollectionViewCell 등록
+    func registerCell() {
+        let nib = UINib(nibName: "ListItemCollectionViewCell", bundle: nil)
+        listCollectionView.register(nib, forCellWithReuseIdentifier: "ListItemCollectionViewCell")
```
[이곳](https://github.com/THE-SOPT-Joint-Seminar-Danggn/Daangn-iOS/pull/7/files#diff-b515d9f2a420da93bf741b6c5d9f8ec8b3f9645a2783e3f6653c01a1e6f51c86R12)에 담인님께서 만들어놓으신 identifier를 사용하시면 문자열 오타 혹은 누락으로 인한 Runtime Error를 피할 수 있을 것 같네요!!

```suggestion
        let nib = UINib(nibName: ListItemCollectionViewCell.identifier, bundle: nil)
        listCollectionView.register(nib, forCellWithReuseIdentifier: ListItemCollectionViewCell.identifier)
```

cchanmi replied:
프로오타러 뜨끔 하고 갑니다

daminoworld replied:
혹시 이곳이 어디를 가르키는 것인가요..?ㅎㅎ

daminoworld replied:
알려주신대로 identifier를 이용해 코드수정하였습니다! 이게 오타를 줄이고 훨씬 좋은것 같네요!

noah0316 replied:
@daminoworld 
> 혹시 이곳이 어디를 가르키는 것인가요..?ㅎㅎ

line number를 link걸어놓았습니다.
바로 화면에 line이 highlighted되지 않는다면 로딩이 끝날 때 까지 3~4초 기다리시는 걸 권장드립니다:)

cchanmi replied:
@noah0316 그런데 세미나 자료에서는 identifier를 셀을 구분하는 id라고 생각하면 된다고 적혀 있는데 identifire를 사용하는 이유가 셀을 구분하는 역할을 함과 동시에 문자열 오타 혹은 누락의 역할도 하는 건가요??

noah0316 commented on Danggn/Screen/Home/Controller/ListItemViewController.swift:
```diff
@@ -8,10 +8,105 @@
+        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        
+            guard let cell = listCollectionView.dequeueReusableCell(
+            withReuseIdentifier: "ListItemCollectionViewCell",
```
위의 리뷰와 동일합니다!

```suggestion
            withReuseIdentifier: ListItemCollectionViewCell.identifier,
```

noah0316 reviewed (commented):
> 네비게이션 바 구현 및 간격 조절을 위해 코드로 구현했는데 더 효율적으로 짤 수 있는 방법이 있는지 궁금합니다..ㅎ
아니면 역시 네비게이션 바 hidden 하고 UIVIew로 구현하는게 더 효율적일까요?

저도 아직 경험이 많이 없어서 잘 모르지만 복잡한 커스텀 네비게이션바가 아니면 이번 PR에서 하신대로 구현하는 것도 좋은 것 같습니다!! 😀

cchanmi commented:
고생하셨습니다 담임쌤 ~~!
네비게이션을 스보로 구현을 하는 것이 생각보다 까다로운 일이라고 해서 옆조분들에게도 여쭤봤는데
다른 기능은 스보로 하였는데 네비게이션은 코드로 구현을 하셨다 하더라고요?
스보로 다루는 법에 대해서는 저도 좀 더 찾아보면서 공부를 해야 할 것 같습니다 ㅠㅠ

noah0316 commented on Danggn/Screen/Home/Model/ListDataModel.swift:
```diff
@@ -0,0 +1,57 @@
+}
+
+extension ListDataModel {
+    static let samleData: [ListDataModel] = [
```
@daminoworld 
`sampleData`
오타 발견!!

noah0316 replied:
@daminoworld 
```suggestion
    static let sampleData: [ListDataModel] = [
```
오타발견!! ㅎㅎ

daminoworld committed:
[Chore] 오타 및 네이밍 관련 코드 수정

- setRightButtonUI 함수 target 인자 제거 -> UI만 설정해주는 함수에서 의미 없는 인자
- listItemBorderView.layer.maskedCorners 에서 위아래 모두 모서리 처리 되도록 수정
- navigationBar -> navigationBarAppearance 수정
- cell identifier 명을 문자열이 아닌 변수로 사용하도록 수정

daminoworld committed:
Merge branch 'feature/1-list-item' of https://github.com/THE-SOPT-Joint-Seminar-Danggn/Daangn-iOS into feature/1-list-item

daminoworld committed:
[Chore] Pods 폴더 제외

- Pods폴더 깃에서 제외 하기 위해 제거 후 커밋

daminoworld committed:
[Chore] git ignore에 Pods 추가

daminoworld committed:
Merge branch 'develop' into feature/1-list-item

noah0316 reviewed (approved):
네트워크 작업 들어가있는거는 다음 PR에서 같이 봐요~
